selfdrive/ui/qt/widgets/cameraview.cc

```diff
@@ -0,0 +1,231 @@
+  connect(timer, &QTimer::timeout, this, &CameraViewWidget::updateFrame);
+}
+
+CameraViewWidget::~CameraViewWidget() {
+  makeCurrent();
+  doneCurrent();
+  glDeleteVertexArrays(1, &frame_vao);
+  glDeleteBuffers(1, &frame_vbo);
+  glDeleteBuffers(1, &frame_ibo);
+}
+
+void CameraViewWidget::initializeGL() {
+  initializeOpenGLFunctions();
+
+  video_rect = {bdr_s, bdr_s, vwp_w - 2 * bdr_s, vwp_h - 2 * bdr_s};
+  gl_shader = std::make_unique<GLShader>(frame_vertex_shader, frame_fragment_shader);
+  GLint frame_pos_loc = glGetAttribLocation(gl_shader->prog, "aPosition");
+  GLint frame_texcoord_loc = glGetAttribLocation(gl_shader->prog, "aTexCoord");
+
+  auto [x1, x2, y1, y2] = stream_type == VISION_STREAM_RGB_FRONT ? std::tuple(0.f, 1.f, 1.f, 0.f) : std::tuple(1.f, 0.f, 1.f, 0.f);
+  const uint8_t frame_indicies[] = {0, 1, 2, 0, 2, 3};
+  const float frame_coords[4][4] = {
+    {-1.0, -1.0, x2, y1}, //bl
+    {-1.0,  1.0, x2, y2}, //tl
+    { 1.0,  1.0, x1, y2}, //tr
+    { 1.0, -1.0, x1, y1}, //br
+  };
+
+  glGenVertexArrays(1, &frame_vao);
+  glBindVertexArray(frame_vao);
+  glGenBuffers(1, &frame_vbo);
+  glBindBuffer(GL_ARRAY_BUFFER, frame_vbo);
+  glBufferData(GL_ARRAY_BUFFER, sizeof(frame_coords), frame_coords, GL_STATIC_DRAW);
+  glEnableVertexAttribArray(frame_pos_loc);
+  glVertexAttribPointer(frame_pos_loc, 2, GL_FLOAT, GL_FALSE,
+                        sizeof(frame_coords[0]), (const void *)0);
+  glEnableVertexAttribArray(frame_texcoord_loc);
+  glVertexAttribPointer(frame_texcoord_loc, 2, GL_FLOAT, GL_FALSE,
+                        sizeof(frame_coords[0]), (const void *)(sizeof(float) * 2));
+  glGenBuffers(1, &frame_ibo);
+  glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, frame_ibo);
+  glBufferData(GL_ELEMENT_ARRAY_BUFFER, sizeof(frame_indicies), frame_indicies, GL_STATIC_DRAW);
+  glBindBuffer(GL_ARRAY_BUFFER, 0);
+  glBindVertexArray(0);
+
+  if (stream_type == VISION_STREAM_RGB_FRONT) {
+    frame_mat = matmul(device_transform, get_driver_view_transform());
+  } else {
+    auto intrinsic_matrix = stream_type == VISION_STREAM_RGB_WIDE ? ecam_intrinsic_matrix : fcam_intrinsic_matrix;
+    float zoom_ = zoom / intrinsic_matrix.v[0];
+    if (stream_type == VISION_STREAM_RGB_WIDE) {
+      zoom_ *= 0.5;
+    }
+    float zx = zoom_ * 2 * intrinsic_matrix.v[2] / video_rect.width();
+    float zy = zoom_ * 2 * intrinsic_matrix.v[5] / video_rect.height();
+
+    const mat4 frame_transform = {{
+      zx, 0.0, 0.0, 0.0,
+      0.0, zy, 0.0, -y_offset / video_rect.height() * 2,
+      0.0, 0.0, 1.0, 0.0,
+      0.0, 0.0, 0.0, 1.0,
+    }};
+    frame_mat = matmul(device_transform, frame_transform);
+  }
+  vipc_client = std::make_unique<VisionIpcClient>("camerad", stream_type, true);
+  timer->start(0);
+}
+
+void CameraViewWidget::resizeGL(int w, int h) {
+  viz_rect = {bdr_s, bdr_s, w - 2 * bdr_s, h - 2 * bdr_s};
+}
+
+void CameraViewWidget::paintGL() {
+  // draw background
+
+  const QColor &color = bg_colors[ui_status];
```

Comment:
the widget shouldn't need a background. the frame should take up the whole widget